### ISSUE TITLE
Cancelling the cancellable doesn't cause sync operations to return

### ISSUE THREAD

**Issue by iainlane** [snapcore/snapd-glib#3]
I'm trying the new timeout facility implemented in gnome-software commit fa7216c1284d957acbf266f879a99ffcff06abee with the snap plugin. On my system at least, the snap plugin often fails to return and this causes gnome-software's operation to never finish (before this commit) or to timeout (with this commit, if the cancellable worked):

```
Thread 10 (Thread 0x7fffc9310700 (LWP 10151)):
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
#1  0x00007ffff74745ff in g_cond_wait (cond=cond@entry=0x55555589ee08, mutex=mutex@entry=0x55555589ee00) at /home/laney/dev/gnome/glib/glib/gthread-posix.c:1395
#2  0x00007ffff741f588 in g_main_context_wait (context=0x55555589ee00, cond=0x55555589ee08, mutex=0x55555589ee00) at /home/laney/dev/gnome/glib/glib/gmain.c:3417
#3  0x00007ffff74207a3 in g_main_loop_run (loop=0x7fffbc0cbe40) at /home/laney/dev/gnome/glib/glib/gmain.c:4140
#4  0x00007fffd2218fec in snapd_client_list_one_sync () from /usr/lib/x86_64-linux-gnu/libsnapd-glib.so.1
#5  0x00007fffd2436388 in gs_plugin_refine_app (plugin=plugin@entry=0x555555aa4590, app=app@entry=0x7fffcc099940, flags=2097184, cancellable=cancellable@entry=0x555557933f60, 
    error=error@entry=0x7fffc930f948) at ../../../../dev/gnome/gnome-software/plugins/snap/gs-plugin-snap.c:451
#6  0x00005555555d7c46 in gs_plugin_loader_call_vfunc (helper=helper@entry=0x5555589466a0, plugin=plugin@entry=0x555555aa4590, app=app@entry=0x7fffcc099940, list=0x5555588ad640, list@entry=0x0, 
    cancellable=cancellable@entry=0x555557933f60, error=error@entry=0x7fffc930fb18) at ../../../../dev/gnome/gnome-software/lib/gs-plugin-loader.c:597
#7  0x00005555555d8b9d in gs_plugin_loader_run_refine_internal (helper=helper@entry=0x5555589466a0, list=list@entry=0x5555588ad640, cancellable=cancellable@entry=0x555557933f60, 
    error=error@entry=0x7fffc930fb18) at ../../../../dev/gnome/gnome-software/lib/gs-plugin-loader.c:806
#8  0x00005555555d93b7 in gs_plugin_loader_run_refine (helper=helper@entry=0x5555588b11e0, list=list@entry=0x5555588ad640, cancellable=cancellable@entry=0x555557933f60, 
    error=error@entry=0x7fffc930fb18) at ../../../../dev/gnome/gnome-software/lib/gs-plugin-loader.c:940
#9  0x00005555555daff8 in gs_plugin_loader_process_thread_cb (task=0x7fffcc0084e0, object=<optimised out>, task_data=0x5555588b11e0, cancellable=0x555557933f60)
    at ../../../../dev/gnome/gnome-software/lib/gs-plugin-loader.c:3150
#10 0x00007ffff70988d9 in g_task_thread_pool_thread (thread_data=0x7fffcc0084e0, pool_data=<optimised out>) at /home/laney/dev/gnome/glib/gio/gtask.c:1328
#11 0x00007ffff7451e6c in g_thread_pool_thread_proxy (data=<optimised out>) at /home/laney/dev/gnome/glib/glib/gthreadpool.c:307
#12 0x00007ffff7451715 in g_thread_proxy (data=0x5555584e6320) at /home/laney/dev/gnome/glib/glib/gthread.c:784
#13 0x00007ffff479d6da in start_thread (arg=0x7fffc9310700) at pthread_create.c:456
#14 0x00007ffff44d717f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:105
```

One of the last `g_debug` log messages I can see is `cancelling job as it took too long`. Let's look at the cancellable that was passed to `snapd-glib`:

```
(gdb) thread 10
[Switching to thread 10 (Thread 0x7fffc9310700 (LWP 10151))]
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
38	in ../sysdeps/unix/sysv/linux/x86_64/syscall.S
(gdb) frame 5
#5  0x00007fffd2436388 in gs_plugin_refine_app (plugin=plugin@entry=0x555555aa4590, app=app@entry=0x7fffcc099940, flags=2097184, cancellable=cancellable@entry=0x555557933f60, 
    error=error@entry=0x7fffc930f948) at ../../../../dev/gnome/gnome-software/plugins/snap/gs-plugin-snap.c:451
451		local_snap = snapd_client_list_one_sync (client, id, cancellable, NULL);
(gdb) print g_cancellable_is_cancelled (cancellable)
$6 = 1
```

so it is cancelled, but the function hasn't returned.

**Comment by robert-ancell** [comment]
When this feature was discussed with Richard I expected we'd need to fix this in snapd-glib, I guess you've confirmed that :)

**Comment by robert-ancell** [comment]
Fixed in fa284cab9090758ce5659a88fec6fbb0e44a8fd8

**Comment by iainlane** [comment]
On 7 July 2017 00:22:06 BST, Robert Ancell <notifications@github.com> wrote:
>Closed #3.

Nice one!

**Comment by iainlane** [comment]
Hmm, no, I don't think this is fully fixed.

See [https://bugs.launchpad.net/ubuntu/+source/gnome-software/+bug/1706715](LP bug #1706715). I was able to reproduce this myself (snapd-glib 1.16).

```
$ gdb -ex run --args gnome-software --verbose
[...]
# editor's picks don't appear
11:03:01:0734 Gs  cancelling job as it took too long
^C
(gdb) t a a bt
[...]
Thread 12 (Thread 0x7fffc5bfd700 (LWP 12426)):
#0  0x00007ffff44b2df9 in syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
#1  0x00007ffff748b75f in g_cond_wait () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#2  0x00007ffff7444a30 in g_main_context_wait () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#3  0x00007ffff7445ab3 in g_main_loop_run () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#4  0x00007fffd00e258c in snapd_client_list_one_sync (client=0x555555a465c0, name=name@entry=0x7fffb8099730 "lxd", cancellable=cancellable@entry=0x555557a82e50, error=error@entry=0x0)
    at snapd-client.c:2809
#5  0x00007fffd030072d in gs_plugin_refine_app (plugin=0x555555a4c1d0, app=0x7fffb809a960, flags=2097184, cancellable=0x555557a82e50, error=0x7fffc5bfc938) at ../plugins/snap/gs-plugin-snap.c:499
#6  0x00005555555c56b9 in gs_plugin_loader_call_vfunc (helper=helper@entry=0x7fffb806f520, plugin=plugin@entry=0x555555a4c1d0, app=app@entry=0x7fffb809a960, list=0x5555588671c0, 
    list@entry=0x0, cancellable=cancellable@entry=0x555557a82e50, error=error@entry=0x7fffc5bfcae8) at ../lib/gs-plugin-loader.c:597
#7  0x00005555555c5da9 in gs_plugin_loader_run_refine_internal (helper=helper@entry=0x7fffb806f520, list=list@entry=0x5555588671c0, cancellable=cancellable@entry=0x555557a82e50, error=error@entry=0x7fffc5bfcae8) at ../lib/gs-plugin-loader.c:807
#8  0x00005555555c65a5 in gs_plugin_loader_run_refine (helper=helper@entry=0x5555587fc640, list=list@entry=0x5555588671c0, cancellable=cancellable@entry=0x555557a82e50, error=error@entry=0x7fffc5bfcae8) at ../lib/gs-plugin-loader.c:955
#9  0x00005555555c70a8 in gs_plugin_loader_process_thread_cb (task=0x7fffdc004e00, object=<optimised out>, task_data=0x5555587fc640, cancellable=0x555557a82e50) at ../lib/gs-plugin-loader.c:3161
#10 0x00007ffff70eab2d in  () at /usr/lib/x86_64-linux-gnu/libgio-2.0.so.0
#11 0x00007ffff746e07e in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#12 0x00007ffff746d685 in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#13 0x00007ffff477f6da in start_thread (arg=0x7fffc5bfd700) at pthread_create.c:456
#14 0x00007ffff44b917f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:105
[...]
(gdb) thread 12
[Switching to thread 12 (Thread 0x7fffc5bfd700 (LWP 12426))]
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
38	../sysdeps/unix/sysv/linux/x86_64/syscall.S: No such file or directory.
(gdb) frame 4
#4  0x00007fffd00e258c in snapd_client_list_one_sync (client=0x555555a465c0, name=name@entry=0x7fffb8099730 "lxd", cancellable=cancellable@entry=0x555557a82e50, error=error@entry=0x0)
    at snapd-client.c:2809
2809	snapd-client.c: No such file or directory.
(gdb) print g_cancellable_is_cancelled (cancellable)
[Thread 0x7fffcec82700 (LWP 12572) exited]
$1 = 1
(gdb) 
```

Hmm, maybe this is the same as #10. If we're hung getting a result back, then we'll be hung processing the cancellable too?
